mlir/lib/Dialect/LLVMIR/IR/GENXDialect.cpp

```diff
@@ -60,8 +66,180 @@ LogicalResult GENXDialect::verifyOperationAttribute(Operation *op,
   return success();
 }
 
-#define GET_OP_CLASSES
-#include "mlir/Dialect/LLVMIR/GENXOps.cpp.inc"
+//===----------------------------------------------------------------------===//
+// Type Parsing
+//===----------------------------------------------------------------------===//
 
-#define GET_ATTRDEF_CLASSES
-#include "mlir/Dialect/LLVMIR/GENXOpsAttributes.cpp.inc"
+static Type parseAndVerifyType(GENXDialect const &dialect,
+                               DialectAsmParser &parser) {
+  Type type;
+  SMLoc typeLoc = parser.getCurrentLocation();
+  if (parser.parseType(type))
+    return Type();
+
+  // Allow GENX dialect types
+  if (&type.getDialect() == &dialect)
+    return type;
+
+  // Check other allowed types
+  if (auto t = type.dyn_cast<FloatType>()) {
+    if (type.isBF16()) {
+      parser.emitError(typeLoc, "cannot use 'bf16' to compose GENX types");
+      return Type();
+    }
+  } else if (auto t = type.dyn_cast<IntegerType>()) {
+    if (!llvm::is_contained({1u, 8u, 16u, 32u, 64u}, t.getWidth())) {
+      parser.emitError(typeLoc,
+                       "only 1/8/16/32/64-bit integer type allowed but found ")
+          << type;
+      return Type();
+    }
+  } else if (auto t = type.dyn_cast<VectorType>()) {
+    if (t.getRank() != 1) {
+      parser.emitError(typeLoc, "only 1-D vector allowed but found ") << t;
+      return Type();
+    }
+    if (t.getNumElements() > 4) {
+      parser.emitError(
+          typeLoc, "vector length has to be less than or equal to 4 but found ")
+          << t.getNumElements();
+      return Type();
+    }
+  } else {
+    parser.emitError(typeLoc, "cannot use ")
+        << type << " to compose GENX types";
+    return Type();
+  }
+
+  return type;
+}
+
+/// Parses the next keyword in `parser` as an enumerator.
+template <typename EnumClass, typename ParserType>
+static ParseResult parseEnumKeywordAttr(EnumClass &value, ParserType &parser,
+                                        StringRef attrName) {
+  StringRef keyword;
+  SmallVector<NamedAttribute, 1> attr;
+  auto loc = parser.getCurrentLocation();
+  if (parser.parseKeyword(&keyword))
+    return failure();
+  if (std::optional<EnumClass> attr = GENX::symbolizeEnum<EnumClass>(keyword)) {
+    value = *attr;
+    return success();
+  }
+  return parser.emitError(loc, "invalid ")
+         << attrName << " attribute specification: " << keyword;
+}
+
+// joint-matrix-type ::= `!genx.jointmatrix` `<`rows `x` cols `x` element-type
+//                                           `,` layout `,` scope`>`
+static Type parseJointMatrixType(GENXDialect const &dialect,
+                                 DialectAsmParser &parser) {
+  if (parser.parseLess())
+    return Type();
+
+  SmallVector<int64_t, 2> dims;
+  SMLoc countLoc = parser.getCurrentLocation();
+  if (parser.parseDimensionList(dims, /*allowDynamic=*/false))
+    return Type();
+
+  if (dims.size() != 2) {
```

Comment:
These restriction are taken from the restriction in the SPIRV specification for a Joint Matrix.

Reply:
Please add comment in code about the source of these restrictions